File: bee/lib/scenegraph/camera.cpp
/*
 * Copyright (c) 2009 Dr. D Studios. (Please refer to license for details)
 * SVN_META_HEADURL = "$HeadURL: http://svn/drd/apps/bee/trunk/lib/scenegraph/camera.cpp $"
 * SVN_META_ID = "$Id: camera.cpp 50498 2010-10-20 22:28:47Z stephane.bertout $"
 */

#include "scenegraph/camera.h"
#include "kernel/assert.h"

#include <OpenEXR/ImathVec.h>

using namespace bee;

fpCameraManagementCallBack Camera::s_CameraManagementCallBack = NULL;

void Camera::setCameraManagementCallBack( fpCameraManagementCallBack a_CameraManagementCallBack )
{
	Assert( s_CameraManagementCallBack == NULL ); // let be sure we store the callback only once
	s_CameraManagementCallBack = a_CameraManagementCallBack;
}

int s_CameraUniqueID = 0;

Camera::Camera( std::string a_Name )
: m_Name( a_Name )
, m_ID( s_CameraUniqueID++ )
, m_ResolutionWidth( 0 )
, m_ResolutionHeight( 0 )
, m_FOV( 90.0f )
, m_NearClipPlane( 0.1f )
, m_FarClipPlane( 10000.0f )
{
	m_Transform.makeIdentity();

	if ( s_CameraManagementCallBack != NULL )
	{
		( *s_CameraManagementCallBack )( this, true ); // add
	}
}

Camera::~Camera()
{
	if ( s_CameraManagementCallBack != NULL )
	{
		( *s_CameraManagementCallBack )( this, true ); // add
	}
}

void Camera::setTransform( const Imath::V3f & a_RotX, const Imath::V3f & a_RotY, const Imath::V3f & a_RotZ, const Imath::V3f & a_Tr )
{
	m_Transform = Matrix( a_RotX.x, a_RotX.y, a_RotX.z, 0.f,
						  a_RotY.x, a_RotY.y, a_RotY.z, 0.f,
						  a_RotZ.x, a_RotZ.y, a_RotZ.z, 0.f,
						  a_Tr.x, a_Tr.y, a_Tr.z, 1.f
						  );
}



/***
    Copyright 2008-2012 Dr D Studios Pty Limited (ACN 127 184 954) (Dr. D Studios)

    This file is part of anim-studio-tools.

    anim-studio-tools is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    anim-studio-tools is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with anim-studio-tools.  If not, see <http://www.gnu.org/licenses/>.
***/

File: grind/python/util/camera.py

import sys
from pimath import *
import math
from . import util # stoopid python thinks it's the standard util one
from OpenGL.GL import *
from OpenGL.GLU import *
from PyQt4 import QtCore


def _Tumble(cam,delta):
    if delta.length() < 0.001:
        return False
    euler = V3f(-delta.y,delta.x,0.0)
    rot = util.GetOrientationMatrix(util.MakeLookAt(cam.lookat,cam.pos,cam.up))
    eulerMat = M44f()
    eulerMat.setEulerAngles(euler)
    eulerMat = util.GetOrientationMatrix(eulerMat)
    xform = rot*(eulerMat.transposed()*rot.transposed())
    pos = cam.lookat + (cam.pos-cam.lookat) * xform
    dir = (cam.lookat-pos).normalized()
    if abs(dir.y) < 0.98:
        cam.pos = pos
    cam.dist = (cam.pos-cam.lookat).length()
    return True

def _Move(cam,delta,staticLA=False):
    if delta.length() < 0.001:
        return False
    delta = V3f(-delta.x,delta.y,delta.z)*(cam.dist*0.4)
    mtx = util.MakeLookAt(cam.pos,cam.lookat,cam.up)
    mtxt = util.GetOrientationMatrix(mtx).transposed()
    offset = V3f(*mtxt.value[0])*delta.x + V3f(*mtxt.value[1])*delta.y + V3f(*mtxt.value[2])*delta.z

    if delta.z == 0:
        cam.pos = cam.pos+offset
        cam.lookat = cam.lookat+offset
    else:
        zD = (1.0+delta.z*3.0)
        zD = zD if zD > 0.0 else 1.0
        v = cam.pos-cam.lookat
        if v.length() < 0.1:
            v = v.normalized() * 0.1
        elif v.length() < 1.5:
             v *= zD
        cam.pos = cam.pos+offset
        #cam.lookat = cam.pos-v
    cam.dist = (cam.pos-cam.lookat).length()

class Camera:
    def __init__(self):
        self.updateViewProj = True
        self.view = M44f()
        self.proj = M44f()
        self.pos = V3f(0,0,1)
        self.up = V3f(0,1,0)
        self.lookat = V3f(0,0,0)
        self.zmin = 0.1
        self.zmax = 100.0
        self.tumbleSpeed = 0.1 / 60.0
        self.mouseMoveSpeed = 0.3 / 60.0
        self.mouseScrollMoveSpeed = 1.5 / 60.0
        self.mouseButtonPressed = None
        self.oldMousePos = V3f(0,0,0)
        self.dist = (self.pos-self.lookat).length()
        self.aspect = 1.33
        self.haperture = 24.892
        self.vaperture = self.haperture / self.aspect
        self.setMM(35.0)

    def setMM(self,mm):
        self.fov = math.atan2((self.haperture/2.0),mm)*2.0
        self.fovY = math.atan2((self.vaperture/2.0),mm)*2.0

    def getMM(self):
        return (self.haperture/2.0) / math.tan(self.fov/2.0)

    def distanceNeeded(self,size):
        h = (size/2.0) / math.tan(self.fov/2.0)
        v = (size/2.0) / math.tan(self.fovY/2.0)
        return max(h,v)

    def mouseButton(self, button, pressed, x, y):
        self.mouseButtonPressed = ( button if pressed else None )
        self.oldMousePos = V3f(x,y,0)

    def mouseMotion(self, x, y):
        delta = V3f(x,y,0)-self.oldMousePos
        tumbleAmount = V3f(0,0,0)
        moveAmount = V3f(0,0,0)
        if self.mouseButtonPressed == QtCore.Qt.LeftButton:
            tumbleAmount += delta*self.tumbleSpeed
        elif self.mouseButtonPressed == QtCore.Qt.MidButton:
            moveAmount.x = delta.x*self.mouseMoveSpeed
            moveAmount.y = delta.y*self.mouseMoveSpeed
        elif self.mouseButtonPressed == QtCore.Qt.RightButton:
            moveAmount.z -= delta.x*self.mouseMoveSpeed
            moveAmount.z += delta.y*self.mouseMoveSpeed

        _Tumble(self,tumbleAmount)
        _Move(self,moveAmount)
        self.oldMousePos = V3f(x,y,0)

    def update(self,elapsed=1.0/60.0):
        if self.updateViewProj:
            self.view = util.MakeLookAt(self.pos,self.lookat,self.up)
            self.proj = util.MakeProjection(self.fov,self.aspect,self.zmin,self.zmax)
        glMatrixMode( GL_PROJECTION );
        glLoadMatrixf( self.proj.value );
        glMatrixMode( GL_MODELVIEW );
        glLoadMatrixf( self.view.value );
        #dist = (self.pos-self.lookat).length()
        #print "p: %5.2f, %5.2f, %5.2f ; la: %5.2f, %5.2f, %5.2f ; u: %5.2f, %5.2f, %5.2f ; D: %5.2f" % (self.pos.x,
        #    self.pos.y, self.pos.z, self.lookat.x, self.lookat.y, self.lookat.z, self.up.x, self.up.y, self.up.z, dist)

    def frame(self,center,height):
        dir = (self.pos-self.lookat).normalized()
        dist = (self.pos-self.lookat).length()
        self.lookat = center
        self.pos = center + dir*self.distanceNeeded(height)*1.4




# Copyright 2008-2012 Dr D Studios Pty Limited (ACN 127 184 954) (Dr. D Studios)
#
# This file is part of anim-studio-tools.
#
# anim-studio-tools is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# anim-studio-tools is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with anim-studio-tools.  If not, see <http://www.gnu.org/licenses/>.

